### PR TITLE
[MERGE] project: improve burndown chart performance

### PR DESCRIPTION
**Pull request by IT-Ideas**
Purpose of this PR
===============

The Burndown Chart report was very slow on big databases as it was not
possible for Postgresql to optimize the query as it was based on a view
that was using several generate series.

This commit aims to improve the performance by injecting the constraints
at a lower level than it was in the past, lowering the amount of data
processed in the higher level of the query.

/!\ Important note
------------------

Overwriting the `read_group_raw` is really not a good practice and should
be avoided in most case. If you fall on this implementation by grepping
the source code, please be advised that this is not the right way of doing
things.

Implementation details
----------------------

- The report is now run by generating the `SQL` that is executed by the
  `read_group_raw`. This allows inserting `SQL` constraints at a lower
  level and simnifically improves performance. As there is no other way
  to do it, the code is unfortunately a modified copy of the actual
  `read_group_raw`.
- The pivot view has been removed as it had no meaning and was creating
  confusing data.
- The `Group By` menu has been limited to `stage_id` and `date` as bringing
  more data trough the different `GROUP BY` statements up to the higher level
  is costly. Further more, additional `Group By` did not bring added value
  as the Chart was less readable.
- The JS code has been adapted in order to force a group by both `stage_id`
  and `date` so that the date displayed is always making sense.
- The sort ascending and descending options have been removed as creating
  confusing data.
- The compare with previous period has also been removed as the chart only
  really make sense when seen chronologically.
- A lot of tests have been added in order to ensure that changes that would
  be harmful for the report will trigger test fails.

task-2845729

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/project/report/project_task_burndown_chart_report.py [hunk 1]
```diff
@@ -3,206 +3,453 @@
...
+    def init(self):
+        # As the ORM requires a table to be created on the models, we need to create a dummy view even if everything
+        # is handled in the read_group_raw by using a custom `SQL`.
         tools.drop_view_if_exists(self.env.cr, self._table)
         self.env.cr.execute(
             sql.SQL("CREATE or REPLACE VIEW {} as ({})").format(
-                sql.Identifier(self._table),
-                sql.SQL(query)
+                    sql.Identifier(self._table),
+                    sql.SQL("SELECT 'dummy_view'")
```
Comment (IT-Ideas) [inline addons/project/report/project_task_burndown_chart_report.py:453]: Remove that useless indent...

**Comment by tivisse** [comment]
@robodoo r+ merge

File: addons/project/report/project_task_burndown_chart_report.py [hunk 2]
```diff
@@ -3,206 +3,453 @@
...
+            # in a sane way
+            result = self._read_group_fill_results(
+                domain, groupby_fields[0], groupby[len(annotated_groupbys):],
+                aggregated_fields, count_field, result, read_group_order=order,
+            )
+        return result
+
+    def init(self):
+        # As the ORM requires a table to be created on the models, we need to create a dummy view even if everything
+        # is handled in the read_group_raw by using a custom `SQL`.
```
Comment (KangOl) [inline addons/project/report/project_task_burndown_chart_report.py:448]: Shouldn't the model be abstract then?

**Comment by tivisse** [comment]
@robodoo r- while waiting a final discussion with @ged-odoo about the control panel.

**Comment by tivisse** [comment]
@robodoo r+ merge

**Pull request merged**